### PR TITLE
add protolock to ci build

### LINKED ISSUE #3368: build: evaluate using protolock in API build
See https://github.com/nilslice/protolock. Should be able to help us codify avoiding accidental breaking changes. The one thing that comes to mind is we are likely going to need some type of mixed mode where certain messages are locked and others are not locked while they are draft. Not sure if that is supported but probably pretty easy to do via some type of comment annotation.

cc @nilslice @rodaine @twoism @htuch

### PR DESCRIPTION
**Pull request by nilslice**
*ci*: *add protolock tool to build step, checking envoy/api protos*

*Description*:
>What does this PR do? What was the behavior before the PR?

Add [`protolock`](https://github.com/nilslice/protolock) lock file to `api` directory and checks `api` protos for various breaking changes. Additionally, running protolock uncovered some syntax errors in a few of the validation options. 

Currently, the `v2alpha` path is the only set of protos ignored. This can be augmented for the case of experimental proto elements by marking them individually with a comment hint, i.e:
```protobuf
// @protolock:skip
message NextLevel_v3 {
// ...
}
```

*Risk Level*: Low 

*Testing*:
>Explanation of what testing was done, for example: unit test,
integration, manual testing, etc.

TODO: push a new branch with an example of an API breaking change within a locked proto. Once the tests run on CI, `protolock status --ignore=v2alpha` should report a warning and stop the build.

Fixes #3368

### CODE REVIEW

File: ci/do_ci.sh [hunk 1]
```diff
@@ -130,12 +130,23 @@ elif [[ "$1" == "bazel.ipv6_tests" ]]; then
...
+  ./protolock status --ignore=v2alpha
+
   cd "${ENVOY_CI_DIR}"
   echo "Building API..."
   bazel --batch build ${BAZEL_BUILD_OPTIONS} -c fastbuild @envoy_api//envoy/...
   echo "Testing API..."
   bazel --batch test ${BAZEL_TEST_OPTIONS} -c fastbuild @envoy_api//test/... @envoy_api//tools/... \
     @envoy_api//tools:capture2pcap_test
+
+  protolock status --ignore=v2alpha
```
Comment (htuch) [inline ci/do_ci.sh]: Why do we run this twice (also run above)?

File: ci/do_ci.sh [hunk 2]
```diff
@@ -130,12 +130,23 @@ elif [[ "$1" == "bazel.ipv6_tests" ]]; then
   exit 0
 elif [[ "$1" == "bazel.api" ]]; then
   setup_clang_toolchain
+ 
+  cd "${ENVOY_CI_DIR}"
+  cd ../api
+  curl https://github.com/nilslice/protolock/releases/download/0.6.0/protolock.20180621T191216Z.linux-amd64.tgz > protolock.tar.gz
```
Comment (htuch) [inline ci/do_ci.sh]: Can you do a SHA verification here as well and pin to a specific SHA? Thanks.

**Comment by nilslice** [comment]
@htuch just wanted to ACK your comments -- will reply and update PR asap.. swamped with some other items at the moment. Thank you for the feedback.

**Comment by nilslice** [comment]
Thanks stalebot. I don't have a precise update at the moment, other than some outside work is blocking my time to get this updated :) Will get back to it ASAP though.

**Pull request closed without merging**